src/main/java/org/jenkinsci/plugins/oic/OicServerWellKnownConfiguration.java

```diff
@@ -116,97 +71,104 @@ public String getWellKnownOpenIDConfigurationUrl() {
-            this.issuer = config.getIssuer();
-            this.tokenServerUrl = config.getTokenEndpoint();
-            this.jwksServerUrl = config.getJwksUri();
-            this.tokenAuthMethod = config.getPreferredTokenAuthMethod();
-            this.userInfoServerUrl = config.getUserinfoEndpoint();
-            if (config.getScopesSupported() != null
-                    && !config.getScopesSupported().isEmpty()) {
-                this.scopes = StringUtils.join(config.getScopesSupported(), " ");
+            OIDCProviderMetadata _oidcProviderMetadata =
+                    OIDCProviderMetadata.parse(rr.retrieveResource(new URL(wellKnownOpenIDConfigurationUrl))
+                            .getContent());
+            String _scopesOverride = getScopesOverride();
+            if (_scopesOverride != null) {
+                // split the scopes by space
+                String[] splitScopes = _scopesOverride.split("\\s+");
+                _oidcProviderMetadata.setScopes(new Scope(splitScopes));
             }
-            this.endSessionUrl = config.getEndSessionEndpoint();
-
-            if (config.getGrantTypesSupported() != null) {
-                this.useRefreshTokens = config.getGrantTypesSupported().contains("refresh_token");
-            } else {
-                this.useRefreshTokens = false;
+            // we do not expose enough to be able to configure all authentication methods,
+            // so limit supported auth methods to CLIENT_SECRET_BASIC / CLIENT_SECRET_POST
+            List<ClientAuthenticationMethod> tokenEndpointAuthMethods =
+                    _oidcProviderMetadata.getTokenEndpointAuthMethods();
+            if (tokenEndpointAuthMethods != null) {
+                List<ClientAuthenticationMethod> filteredEndpointAuthMethods =
+                        new ArrayList<>(tokenEndpointAuthMethods);
+                filteredEndpointAuthMethods.removeIf(cam -> cam != ClientAuthenticationMethod.CLIENT_SECRET_BASIC
+                        && cam != ClientAuthenticationMethod.CLIENT_SECRET_POST);
+                if (filteredEndpointAuthMethods.isEmpty()) {
+                    LOGGER.log(
+                            Level.WARNING,
+                            "OIDC well-known configuration reports only unsupported token authentication methods (authentication may not work): "
+                                    + tokenEndpointAuthMethods.stream()
+                                            .map(Object::toString)
+                                            .collect(Collectors.joining(",", "[", "]")));
+                    _oidcProviderMetadata.setTokenEndpointAuthMethods(null);
+                } else {
+                    _oidcProviderMetadata.setTokenEndpointAuthMethods(filteredEndpointAuthMethods);
+                }
             }
-
-            setWellKnownExpires(response.getHeaders());
+            // do not allow the "none" singing algorithm for security
```

Comment:
Whilst we are using code flow, it is not needed, but providers MUST support RS256 so there would always be a more secure option we can use.